keps/sig-auth/4412-projected-service-account-tokens-for-kubelet-image-credential-providers/README.md

```diff
@@ -0,0 +1,1079 @@
++
+        // GlobalPluginCacheKeyType means the kubelet will cache credentials for all images that
+        // match for a given plugin. This cache key should only be returned by plugins that do not use
+        // the image input at all.
+        GlobalPluginCacheKeyType PluginCacheKeyType = "Global"
+ )
+ 
++type PluginCacheMode string
++
++const (
++       // ServiceAccountPluginCacheMode means the kubelet will cache credentials on a per-service account basis in addition to the cache key type.
++       // This cache mode should be used when the plugin wants to cache credentials per service account for the image or registry.
++       // The cache key will be the service account info + cache key (image or registry).
++       // This cache mode has no effect if the cache key type is Global.
++       ServiceAccountPluginCacheMode PluginCacheMode = "ServiceAccount"
++
++       // NonePluginCacheMode means the kubelet will cache credentials based on the cache key type only (default behavior).
++       NonePluginCacheMode PluginCacheMode = "None"
++)
++
+ // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+ 
+ // CredentialProviderResponse holds credentials that the kubelet should use for the specified
+@@ -63,6 +96,12 @@ type CredentialProviderResponse struct {
+        // Global. If an invalid value is specified, the response will NOT be used by the kubelet.
+        CacheKeyType PluginCacheKeyType
+ 
++       // cacheMode indicates the mode of caching to use based on the cache key type.
++       // This field is required if the audience field is configured in the kubelet's credential provider configuration.
++       // There are two valid values for the cache mode: ServiceAccount and None.
++       // If an invalid value is specified, the response will NOT be used by the kubelet.
++       CacheMode PluginCacheMode
+```
+
+#### Caching Credentials
+
+The cache key generation will be updated to support new caching modes while preserving the existing behavior when the audience field is not set.
+
+##### Existing behavior
+
+If the `audience` field is not set in the provider configuration for the plugin, the key generation should fall back to the current behavior: using only the image registry URL or other existing identifiers.
+
+##### New behavior when the `audience` field is set
+
+Based on the `PluginCacheKeyType` definitions (Image, Registry, Global) and the newly introduced cache modes (ServiceAccount, None), here’s a breakdown of how cache keys would be generated and whether cache mode impacts their behavior:
+
+1. `GlobalPluginCacheKeyType` (Global)
+   - cache mode does not affect the cache key generation since it is designed to provide credentials for all images.
+2. `RegistryPluginCacheKeyType` (Registry)
+   - `ServiceAccount` cache mode:
+     - The cache key will be generated based on the image registry URL, service account token bound to the pod (namespace, name, UID), hash of the pod annotations.
```

Comment:
pod annotations are questionable as a cache key... admission and controllers put weird random values in those that differ across replicas

service account annotations would be significantly more stable (and would map much more naturally to the service account being the partitioning dimension), though I don't love pushing opaque user-controlled values down to providers

making the credential provider opt into getting the annotations (maybe even the specific annotations it wants) avoids sending arbitrary annotation content down to it (including stuff like client-side apply annotations), and shrinks the set of annotations that could invalidate caching if they change